server/aesdsocket.c

```diff
@@ -0,0 +1,161 @@
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM; 
+
+    openlog(NULL, 0, LOG_USER);
+
+    signal(SIGINT, signal_handler);  //set up the signal handler
+    signal(SIGTERM, signal_handler);  
+
+//getaddrinfo provides the socket address
+    if ((status = getaddrinfo(NULL, PORT, &hints, &servinfo)) != 0) {
+        syslog(LOG_ERR, "getaddrinfo failed: %s", gai_strerror(status));
+        return -1;
+    }
+//create a socket for communication
+    sockfd = socket(servinfo->ai_family, servinfo->ai_socktype, servinfo->ai_protocol);
+    if (sockfd == -1) {
+        syslog(LOG_ERR, "Socket creation failed");
+        freeaddrinfo(servinfo);
+        return -1;
+    }
+//bind the socket to an address and port number   
+    if (bind(sockfd, servinfo->ai_addr, servinfo->ai_addrlen) == -1) {
+        syslog(LOG_ERR, "Bind failed");
+        close(sockfd);
+        freeaddrinfo(servinfo);
+        return -1;
+    }
+    freeaddrinfo(servinfo);
+//server waits for client to make connection
+    if (listen(sockfd, BACKLOG) == -1) {
+        syslog(LOG_ERR, "Listen failed");
+        close(sockfd);
+        return -1;
+    }
+//go to daemon mode after binding to the port
+    if (argc > 1 && strcmp(argv[1], "-d") == 0) {  //checking if -d arg is passed
+        daemonize(); //call daemon function
+    }
+//loop until connection is made and data is transfered without any interrupts
+    while (1) {
+        new_fd = accept(sockfd, (struct sockaddr *)&client_addr, &addr_size); //accepts connection with a requested node
+        if (new_fd == -1) {  //client node
+            syslog(LOG_ERR, "Accept failed");
+            continue;  
+        }
+//extract the IP address of the client
+        inet_ntop(client_addr.ss_family, get_in_addr((struct sockaddr *)&client_addr), s, sizeof s);
+        syslog(LOG_INFO, "Accepted connection from %s", s);
+       
+//open file to write and read contents
+     FILE *fd = fopen(write_file, "a+"); //open the file at the specied path/ create if it doesnt exist
+    if (fd == NULL) {
+         syslog(LOG_ERR, "File open failed");
+          close(new_fd);
+          return -1; 
+    }
+//receive data from the client node
+        while ((bytes_received = recv(new_fd, buffer, BUFFER_SIZE, 0)) > 0) {
+           buffer[bytes_received] = '\0'; //null terminate the packets
+           fwrite(buffer,1,bytes_received,fd); //block transfer of the received bytes to the file 
+           fflush(fd); //flush immediately
+           fseek(fd, 0, SEEK_SET);  // Reset file pointer to the begining 
+           if (strstr(buffer, "\n"))  //if newline is found in the data stream
+           { while ((fgets(read_buffer, BUFFER_SIZE, fd))>0) { //read the data line by line from the file
+                size_t data_length = strlen(read_buffer);
+                    send(new_fd,read_buffer, data_length, 0); //send data back to the client 
+                }
+            } 
+        }
+        fclose(fd); //close the file
```

Comment:
You won't do this cleanup when you get a signal and exit() from there, this is what your valgrind log is telling you (that you didn't close this file descriptor).